heron/api/src/java/org/apache/heron/streamlet/Streamlet.java

```diff
@@ -71,6 +71,22 @@
    */
   int getNumPartitions();
 
+  /**
+   * Set the id of the stream to be used by the children nodes.
+   * Usage (assuming source is a Streamlet object with two output streams: stream1 and stream2):
+   *   source.withStream("stream1").filter(...).log();
+   *   source.withStream("stream2").filter(...).log();
+   * @param streamId The specified stream id
+   * @return Returns back the Streamlet with changed stream id
+   */
+  Streamlet<R> withStream(String streamId);
+
+  /**
+   * Gets the stream id of this Streamlet.
+   * @return the stream id of this Streamlet
+   */
+  String getStreamId();
```

Comment:
Also IT coverage will be useful to verify end2end so will be addressing in the light of #3106

Reply:
In fact, I am thinking of adding the split() operator now with this PR so that we can add a proper IT for it. I prefer to break PR down to single functionality but in this case it could be useful.

Reply:
Sure, IT coverage is too important due to aim end2end. I think it should be part of the code review checklist for incoming PRs ;) 

On the other hand, do you have template for `split` function signature?

Reply:
Did a quick search, I think it is pretty close to the split operator in Flink (https://ci.apache.org/projects/flink/flink-docs-release-1.6/dev/stream/operators/).

In fact, select() seems to be a better name than "withStream". :D I will think about it.

Neng gives a feedback about the withStream() function as a new user and he made a good point that we need to handle both cases:
1.
  stream.map();
  stream.withStream("a").map();
  stream.map();   // this should be the default stream
  stream.withStream("a").map();

2.
  steam2 = stream.withString("a");
  stream2.map();   // this should be the new stream
  stream.map();   // this should be the default stream

And this is not supported with the current implementation since the this object is muted. I am going to see what is the best way and update this PR.

Reply:
Yep, i was having a look for Flink implementation as well. It supports child and parent Stream hierharchy to support `select` functions through `split` function.

I think more make sense `split` function naming convention and following can be as an alternative: 

```
Streamlet<Integer> numberStreamlets = ...
// user can apply his custom split logic
Map<String, Streamlet> splittedStreamlets = numberStreamlets.split(new SerializableSplitter())

splittedStreamlets.get("even")
splittedStreamlets.get("odd")
```

If we have proposal document by covering the all potential use-cases, that will be useful as well ;)

Reply:
"Map<String, Streamlet<Integer>>"  won't really work because right now one Streamlet is mapped to one operator which is a Bolt component. In our case, there is only one component (hence one streamlet object) with multiple streams.

I had to add a StreamletShadow class to create companion objects of the main Streamlet object in my implementation.

Reply:
In fact, still no....

The problem is the .setName and .setPartition functions, like:

numberStreamlets.split(splitter).setName("split").setPartition(10);

the split operator is one single bolt and all entries in Map<String, Streamlet> points to the same bolt. It might be ok to set name and partition like:
Map<String, Streamlet> splittedStreamlets  = numberStreamlets.split(splitter);

splittedStreamlets.get("even").setName("split")
splittedStreamlets.get("even").setPartition(partition)

but it looks really bad this way.